### PR TITLE
ENH: _BaseReader can be used as context guard.

### PR DESCRIPTION
**Pull request by nme01**
Now if someone creates a lot of requests, he can use Readers using the `with` statement so that the session is correctly closed, e.g.:
```python
with YahooDailyReader(symbols, start=start, end=end) as reader:
    return reader.read()
```

### CODE REVIEW

**Comment by femtotrader** [comment]
Maybe you should add a test case (or modify an existing one) to ensure that this new feature works as expected.

**Comment by nme01** [comment]
I'm not exactly sure how this should be done. I would need to start a connection with something (e.g. using session's mock). However `_BaseReader` is incapable of reading anything from a session (even if I'd used session's mock) since its `url` property is abstract (raises `NotImplementedError`). Could you give me a hand with this?

**Comment by bashtage** [comment]
You could test using anything that derives from _BaseReader that should
inherit the context manager.

**Comment by bashtage** [comment]
Is there a good reason that `read()` doesn't always close?  If it was implemented this way there would not be a need for a CM.

**Comment by femtotrader** [comment]
Maybe because you can use same `session` for several readings.

**Comment by bashtage** [comment]
That was my question.  Is this a realistic use case and is there any gain in allowing session reuse?

**Comment by femtotrader** [comment]
I wonder why we should recreate an other `Session` (or `CachedSession`) instance when parameters to build them are the same?  Is there any gain in instantiating again a `Session` (or `CachedSession`?
Most of us never close `session` when using for example IPython. So my question would be why closing `session` would be required. I think it's not necessary to do it (we never did it previously) and it didn't hurt. So the idea of using a context manager is nice as it doesn't urge people to use it.
But I won't use this CM is `get_data_fred`, `get_data_famafrench` ... because it will close session by default.
I will only use this CM in one test of a DataReader that derives from `_BaseReader`

**Comment by nme01** [comment]
However when you use `get_data_fred`, `get_data_famafrench`, `...` you create a new session on each call. There is no option to pass an already opened session. It doesn't even return the created session. I created the pull request, because we got a lot of `ResourceWarnings` when we ran our tests with `unittests` each time the read failed (e.g. because it couldn't reach YahooFinance). It should be ensured that the session is properly closed in these cases.

**Comment by femtotrader** [comment]
Quote from: @nme01 

    However when you use get_data_fred, get_data_famafrench, ... you create a new session on each call. There is no option to pass an already opened session. It doesn't even return the created session. I created the pull request, because we got a lot of ResourceWarnings when we ran our tests with unittests each time the read failed (e.g. because it couldn't reach YahooFinance). It should be ensured that the session is properly closed in these cases.

It seems that `get_data_fred`, `get_data_famafrench` accept keyword args and `session` is one.
You don't need to return the "created session", session is an object on which you can keep a reference.

**Comment by nme01** [comment]
I realised that right after I posted my comment. That's why I've deleted it.

File: pandas_datareader/base.py [hunk 1]
```diff
@@ -55,6 +55,12 @@ def __init__(self, symbols, start=None, end=None,
         self.timeout = timeout
         self.session = _init_session(session, retry_count)
```
Comment (bashtage) [inline pandas_datareader/base.py]: You should keep track of whether it was a user session or a new session.

File: pandas_datareader/base.py [hunk 2]
```diff
@@ -55,6 +55,12 @@ def __init__(self, symbols, start=None, end=None,
...
 
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exception_type, exception_value, exception_traceback):
+        self.session.close()
```
Comment (bashtage) [inline pandas_datareader/base.py]: Only close if it was a new session.

File: pandas_datareader/data.py [hunk 3]
```diff
@@ -25,35 +25,43 @@
...
 
 def get_data_google(*args, **kwargs):
-    return GoogleDailyReader(*args, **kwargs).read()
+    with GoogleDailyReader(*args, **kwargs) as reader:
+        return reader.read()
 
 
 def get_data_yahoo(*args, **kwargs):
-    return YahooDailyReader(*args, **kwargs).read()
+    with YahooDailyReader(*args, **kwargs) as reader:
```
Comment (bashtage) [inline pandas_datareader/data.py]: Don't need to test all readers.  Lets  stick to Fred and FamaFrench since these are likely to work.

**Comment by bashtage** [comment]
I'm going to close this since I don't see a compelling case.

**Pull request closed without merging**